File: topological-sort/topological_sort.go
package main
import "fmt"

func dfs(node int, graph map[int][]int, visited map[int]bool, stack *[]int) {
    visited[node] = true
    
    for _, neighbor := range graph[node] {
        if !visited[neighbor] {
            dfs(neighbor, graph, visited, stack)
        }
    }
    
    *stack = append(*stack, node)
}

func topologicalSort(graph map[int][]int) []int {
    visited := make(map[int]bool)
    var stack []int
    
    for node := range graph {
        if !visited[node] {
            dfs(node, graph, visited, &stack)
        }
    }
    
    // Reverse the stack
    for i, j := 0, len(stack)-1; i < j; i, j = i+1, j-1 {
        stack[i], stack[j] = stack[j], stack[i]
    }
    
    return stack
}

func main() {
    graph := map[int][]int{
        0: {1, 2},
        1: {3},
        2: {3},
        3: {},
    }
    
    result := topologicalSort(graph)
    fmt.Println(result)
}

File: topological-sort/topological_sort.cs
using System;
using System.Collections.Generic;

class TopologicalSort {
    static void Dfs(int node, Dictionary<int, List<int>> graph, HashSet<int> visited, Stack<int> stack) {
        visited.Add(node);
        
        if (graph.ContainsKey(node)) {
            foreach (int neighbor in graph[node]) {
                if (!visited.Contains(neighbor)) {
                    Dfs(neighbor, graph, visited, stack);
                }
            }
        }
        
        stack.Push(node);
    }
    
    static List<int> Sort(Dictionary<int, List<int>> graph) {
        var visited = new HashSet<int>();
        var stack = new Stack<int>();
        
        foreach (int node in graph.Keys) {
            if (!visited.Contains(node)) {
                Dfs(node, graph, visited, stack);
            }
        }
        
        var result = new List<int>();
        while (stack.Count > 0) {
            result.Add(stack.Pop());
        }
        
        return result;
    }
    
    static void Main() {
        var graph = new Dictionary<int, List<int>> {
            {0, new List<int> {1, 2}},
            {1, new List<int> {3}},
            {2, new List<int> {3}},
            {3, new List<int>()}
        };
        
        var result = Sort(graph);
        Console.WriteLine(string.Join(" ", result));
    }
}